File: packages/sherpa-native/android/src/main/java/dev/happier/sherpa/FrameFedVadDetectorRegistry.kt
package dev.happier.sherpa

import android.content.Context
import dev.happier.sherpa.vad.VadModelAssetInstaller

internal data class FrameFedVadFrameResult(
  val speechStarted: Boolean,
  val speechEnded: Boolean
)

internal interface NativeVadDriver {
  fun create(modelPath: String, sampleRate: Int, minSpeechSec: Float, minSilenceSec: Float): Long
  fun push(handle: Long, pcm16: ShortArray, channels: Int): Int
  fun destroy(handle: Long)
}

private object JniNativeVadDriver : NativeVadDriver {
  override fun create(modelPath: String, sampleRate: Int, minSpeechSec: Float, minSilenceSec: Float): Long =
    HappierSherpaNativeJni.nativeCreateVadSession(modelPath, sampleRate, minSpeechSec, minSilenceSec)

  override fun push(handle: Long, pcm16: ShortArray, channels: Int): Int =
    HappierSherpaNativeJni.nativeVadAcceptPcm16(handle, pcm16, pcm16.size, channels)

  override fun destroy(handle: Long) {
    HappierSherpaNativeJni.nativeDestroyVadSession(handle)
  }
}

internal interface FrameFedVadDetectorRegistryControl {
  fun create(detectorId: String, sampleRate: Int, minSpeechMs: Long, redemptionMs: Long)
  fun push(detectorId: String, pcm16: ShortArray, sampleRate: Int, channels: Int): FrameFedVadFrameResult
  fun cancel(detectorId: String)
  fun cancelAll()
}

/** Inference-only VAD ownership. It never opens AudioRecord or changes audio focus/session state. */
internal class FrameFedVadDetectorRegistry(
  context: Context? = null,
  private val driver: NativeVadDriver = JniNativeVadDriver,
  private val resolveModelPath: () -> String = {
    VadModelAssetInstaller.ensureInstalled(requireNotNull(context) { "Android context is required" })
  }
) : FrameFedVadDetectorRegistryControl {
  private data class Entry(
    val handle: Long,
    val sampleRate: Int,
    var speechActive: Boolean = false
  )

  private val entries = mutableMapOf<String, Entry>()

  @Synchronized
  override fun create(detectorId: String, sampleRate: Int, minSpeechMs: Long, redemptionMs: Long) {
    require(detectorId.isNotBlank()) { "detectorId is required" }
    require(sampleRate > 0) { "sampleRate must be positive" }

    val modelPath = resolveModelPath()
    val handle = driver.create(
      modelPath,
      sampleRate,
      (minSpeechMs.coerceAtLeast(0) / 1000.0).toFloat(),
      (redemptionMs.coerceAtLeast(0) / 1000.0).toFloat()
    )
    check(handle != 0L) { "failed_to_create_vad_detector" }

    entries.put(detectorId, Entry(handle, sampleRate))?.let { driver.destroy(it.handle) }
  }

  @Synchronized
  override fun push(
    detectorId: String,
    pcm16: ShortArray,
    sampleRate: Int,
    channels: Int
  ): FrameFedVadFrameResult {
    val entry = entries[detectorId] ?: throw IllegalStateException("vad_detector_not_found")
    require(sampleRate == entry.sampleRate) { "vad_sample_rate_mismatch" }
    require(channels > 0) { "channels must be positive" }

    val state = driver.push(entry.handle, pcm16, channels)
    val speechDetected = state and 1 != 0
    val speechEnded = state and 2 != 0
    val speechStarted = speechDetected && !entry.speechActive
    entry.speechActive = speechDetected && !speechEnded
    return FrameFedVadFrameResult(speechStarted, speechEnded)
  }

  @Synchronized
  override fun cancel(detectorId: String) {
    entries.remove(detectorId)?.let { driver.destroy(it.handle) }
  }

  @Synchronized
  override fun cancelAll() {
    val active = entries.values.toList()
    entries.clear()
    active.forEach { driver.destroy(it.handle) }
  }
}

File: packages/sherpa-native/ios/FrameFedVadDetectorRegistry.swift
import Foundation

struct FrameFedVadDetection {
  let speechDetected: Bool
  let speechEnded: Bool
}

protocol FrameFedVadDetector: AnyObject {
  func acceptPcm16(data: Data, sampleRate: Int32, channels: Int32) throws -> FrameFedVadDetection
  func close()
}

struct FrameFedVadFrameResult {
  let speechStarted: Bool
  let speechEnded: Bool
}

/// Owns inference-only VAD detectors. Microphone and audio-session ownership
/// deliberately remain outside this package in audio-stream-native.
final class FrameFedVadDetectorRegistry {
  typealias Factory = (_ sampleRate: Int32, _ minSpeechSec: Float, _ minSilenceSec: Float) throws -> FrameFedVadDetector

  private final class Entry {
    let detector: FrameFedVadDetector
    var speechActive = false

    init(detector: FrameFedVadDetector) {
      self.detector = detector
    }
  }

  private let lock = NSLock()
  private let makeDetector: Factory
  private var entries: [String: Entry] = [:]

  init(makeDetector: @escaping Factory) {
    self.makeDetector = makeDetector
  }

  func create(detectorId: String, sampleRate: Int32, minSpeechMs: Int64, redemptionMs: Int64) throws {
    let id = detectorId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !id.isEmpty else {
      throw NSError(domain: "HappierSherpaNative", code: 401, userInfo: [NSLocalizedDescriptionKey: "detectorId is required"])
    }
    guard sampleRate > 0 else {
      throw NSError(domain: "HappierSherpaNative", code: 402, userInfo: [NSLocalizedDescriptionKey: "sampleRate must be positive"])
    }

    let detector = try makeDetector(
      sampleRate,
      Float(max(0, minSpeechMs)) / 1000.0,
      Float(max(0, redemptionMs)) / 1000.0
    )

    lock.lock()
    let previous = entries.updateValue(Entry(detector: detector), forKey: id)
    lock.unlock()
    previous?.detector.close()
  }

  func push(detectorId: String, data: Data, sampleRate: Int32, channels: Int32) throws -> FrameFedVadFrameResult {
    lock.lock()
    defer { lock.unlock() }

    guard let entry = entries[detectorId] else {
      throw NSError(domain: "HappierSherpaNative", code: 403, userInfo: [NSLocalizedDescriptionKey: "VAD detector not found"])
    }

    let detection = try entry.detector.acceptPcm16(data: data, sampleRate: sampleRate, channels: channels)
    let speechStarted = detection.speechDetected && !entry.speechActive
    entry.speechActive = detection.speechDetected && !detection.speechEnded
    return FrameFedVadFrameResult(speechStarted: speechStarted, speechEnded: detection.speechEnded)
  }

  func cancel(detectorId: String) {
    lock.lock()
    let entry = entries.removeValue(forKey: detectorId)
    lock.unlock()
    entry?.detector.close()
  }

  func cancelAll() {
    lock.lock()
    let active = Array(entries.values)
    entries.removeAll()
    lock.unlock()
    active.forEach { $0.detector.close() }
  }
}
